<a href="https://colab.research.google.com/github/rdevia/text_mining_unrn/blob/main/finetunning_1123.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Uso de modelos pre-entrenados para clasificación y finetunning


**Natalia Dedandi**


Vamos a usar las librerías y datasets de Hugging face

https://huggingface.co/

In [1]:
!pip install datasets seaborn
! pip install -U accelerate
! pip install -U transformers



1- Crear una cuenta en HUGGINGFACE

https://huggingface.co/

In [2]:
from huggingface_hub import notebook_login
from google.colab import userdata

notebook_login()

In [3]:
pip install -U datasets

In [4]:
from datasets import load_dataset
import pandas as pd
from datetime import datetime, timedelta
from tqdm.auto import tqdm
from collections import Counter

Análisis exploratorio inicial

https://huggingface.co/datasets/piuba-bigdata/contextualized_hate_speech



*   ¿Cómo es este dataset?

*   ¿Que estructura es ds1?¿cuantos conjuntos tiene?






1- carguen el dataset

In [5]:
from datasets import load_dataset

ds1 = load_dataset("piuba-bigdata/contextualized_hate_speech")



/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


In [6]:
ds1

DatasetDict({
    train: Dataset({
        features: ['id', 'title', 'text', 'context_tweet', 'HATEFUL', 'body', 'CALLS', 'WOMEN', 'LGBTI', 'RACISM', 'CLASS', 'POLITICS', 'DISABLED', 'APPEARANCE', 'CRIMINAL'],
        num_rows: 36420
    })
    test: Dataset({
        features: ['id', 'title', 'text', 'context_tweet', 'HATEFUL', 'body', 'CALLS', 'WOMEN', 'LGBTI', 'RACISM', 'CLASS', 'POLITICS', 'DISABLED', 'APPEARANCE', 'CRIMINAL'],
        num_rows: 11343
    })
    dev: Dataset({
        features: ['id', 'title', 'text', 'context_tweet', 'HATEFUL', 'body', 'CALLS', 'WOMEN', 'LGBTI', 'RACISM', 'CLASS', 'POLITICS', 'DISABLED', 'APPEARANCE', 'CRIMINAL'],
        num_rows: 9106
    })
})

El dataset 1 contiene 56869 tw etiquetados.
El dataset se dividió en 36420 de train, test:11343 y vaLidacion:9106

In [7]:
ds1_train=ds1["train"]
ds1_train

Dataset({
    features: ['id', 'title', 'text', 'context_tweet', 'HATEFUL', 'body', 'CALLS', 'WOMEN', 'LGBTI', 'RACISM', 'CLASS', 'POLITICS', 'DISABLED', 'APPEARANCE', 'CRIMINAL'],
    num_rows: 36420
})

1) Dataset de etiquetado

https://huggingface.co/datasets/piuba-bigdata/contextualized_hate_speech

Este es el dataset original de etiquetado. Contiene los siguientes campos:

ID : identificador

TITLE: título de la noticia

TEXT: texto del tweet del comentario al Tweet original

CONTEXT: tweet original

HATEFUL: 1 hateful, 0 not hateful

BODY: texto de la noticia

CALLS: si se realizó o no un llamado a la acción

WOMEN: discriminación hacia mujeres

LGBTI: discriminación hacia grupos LGBTI

RACISM: discriminación asociada a racismo

CLASS: discriminación vinculada a pobreza, villas

POLITICS: odio o violencia vinculada a las ideas políticas

DISABLED: discriminación a las personas con discapacidad

APPEARENCE: discriminación en base al aspecto físico (incluye especialmente la gordofobia)

CRIMINAL: discriminación hacia personas en el sistema penal o similares

**Para el entrenamiento se usaron los siguientes tutoriales y documentos:**

Uso  de base el siguiente tutorial para hacer el finetuning:
https://huggingface.co/blog/sentiment-analysis-python
y este:
https://huggingface.co/docs/transformers/training


La idea es utilizar la librería tranbsformers de huggingface para hacerle finetuning a un BETO original, es decir la version de BERT pre-entrenada con español de GOOGLE pero que no haya sido reentrenado con ningún otro dataset.
Aqui para usar transformers de hugginface
https://github.com/huggingface/transformers?tab=readme-ov-file



In [8]:
import torch
torch.cuda.is_available()

True

### 1. pre-process data

In [9]:
 ds1['train'][0]

{'id': 343726,
 'title': 'Video: salió de la cárcel por el coronavirus y murió de un tiro el mismo día al festejar su libertad',
 'text': '@usuario Uno menos',
 'context_tweet': 'Video: salió de la cárcel por el coronavirus y murió de un tiro el mismo día al festejar su libertad https://t.co/UrJGRAdEvD',
 'HATEFUL': 1,
 'body': 'Un hombre de 46 años, que cumplía una condena en prisión y fue puesto en libertad por el coronavirus, murió el mismo día que dejó la cárcel baleado por un primo, durante los "festejos" por dejar la prisión.\n\nSari Salem Wardat, de 46 años, recibió un disparo desde atrás en un aparente accidente mientras caminaba hacia miembros de la familia y vecinos que se habían reunido para darle la bienvenida a su casa, en Jordania.\n\nUn video que circuló en las redes sociales \u200bmuestra a Sari llegando a su casa en un auto azul rodeado de tiros de "celebración". El primo se acerca de inmediato a Sari, que sostiene una pistola y la apunta hacia el cielo.\n\nEl detenido

In [10]:
##ME quedo solo con las columnas que necesito para la clasificación binaria
ds_train_small= ds1['train'].select_columns(["text","HATEFUL"])
ds_test_small= ds1['test'].select_columns(["text","HATEFUL"])
ds_val_small= ds1['dev'].select_columns(["text","HATEFUL"])

Nos quedamos con las columnas Text y Hateful y se renombra esta última ya que el modelo requiere que se llame "label"

In [11]:
##aca les coloco el mismo nombre que espera el modelo (me parece cualquiera pero no logre que ande sino)
ds_train_small=ds_train_small.rename_column("HATEFUL", "label")
ds_test_small=ds_test_small.rename_column("HATEFUL", "label")
ds_val_small=ds_val_small.rename_column("HATEFUL", "label")

In [12]:
ds_train_small

Dataset({
    features: ['text', 'label'],
    num_rows: 36420
})

In [13]:
ds_test_small

Dataset({
    features: ['text', 'label'],
    num_rows: 11343
})

In [14]:
# Tomo una muestra de 50 mil para que sea más rapido el entrenamiento
#  uso una semilla, para que pueda reproducirlo, es decir para que mi muestra siempre sea la misma.
# cada uno elija su numero

import random

# Set the random seed for reproducibility
random.seed(17)

# Sample 50,000 random examples from ds_test_small
indices_to_sample = random.sample(range(len(ds_test_small)), 10000)
ds_train_small = ds_train_small.select(indices_to_sample)

ds_train_small


Dataset({
    features: ['text', 'label'],
    num_rows: 10000
})

In [15]:
ds_test_small[0]

{'text': '@usuario Ojalá se funda y cague de hambre', 'label': 0}

In [16]:
!pip install --upgrade jax jaxlib

In [17]:
from transformers import BertForMaskedLM, BertTokenizer

/usr/local/lib/python3.11/dist-packages/jaxlib/plugin_support.py:71: RuntimeWarning: JAX plugin jax_cuda12_plugin version 0.5.1 is installed, but it is not compatible with the installed jaxlib version 0.5.3, so it will not be used.
  warnings.warn(


Utiizo las librerías de transformers para preparar los datos de entrenamiento

Lo primero es tokenizar en función del modelo seleccionado, en nuestro caso BETO

In [18]:
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("dccuchile/bert-base-spanish-wwm-uncased")

tokenizer_config.json:   0%|          | 0.00/310 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/650 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/248k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/486k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/134 [00:00<?, ?B/s]

Le aplico la función tokenizer del modelo BETO a mis conjuntos, el de entrenamiento, el de test y el de validación

In [19]:
def preprocess_function(examples):
   return tokenizer(examples["text"], truncation=True)

tokenized_train = ds_train_small.map(preprocess_function, batched=True)
tokenized_test = ds_test_small.map(preprocess_function, batched=True)
tokenized_val = ds_val_small.map(preprocess_function, batched=True)

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

Map:   0%|          | 0/11343 [00:00<?, ? examples/s]

Map:   0%|          | 0/9106 [00:00<?, ? examples/s]

In [20]:
tokenized_train[0]

{'text': '@usuario acaso será el cerdo de bolsonaro?',
 'label': 1,
 'input_ids': [4,
  985,
  8563,
  7291,
  2119,
  1039,
  9160,
  1009,
  12359,
  12184,
  30957,
  1059,
  5],
 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

Prueben varios casos

In [21]:
tokenized_train[2100]

{'text': '@usuario Reclamale a los que votaste jaa',
 'label': 0,
 'input_ids': [4,
  985,
  8563,
  27093,
  1080,
  1012,
  1067,
  1041,
  2107,
  4985,
  2279,
  30956,
  5],
 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

La siguiente función transforma los datos de entrenamiento agregandole el padding

Se usará luego en el Trainer

In [22]:
from transformers import DataCollatorWithPadding
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

## 2 Entreno el modelo.

Voy a entrenar el modelo con el TRAINER de transformers. Instancio el modelo pre-entrenado BETO y aca selecciono 2 labels ya que se hace una clasificación binaria

Modelo: dccuchile/bert-base-spanish-wwm-uncased

Ustedes tendrán que probar otros modelos


In [23]:
from transformers import AutoModelForSequenceClassification
model = AutoModelForSequenceClassification.from_pretrained("dccuchile/bert-base-spanish-wwm-uncased", num_labels=2)


pytorch_model.bin:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at dccuchile/bert-base-spanish-wwm-uncased and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [24]:
#esto es opcional
#import wandb
#wandb.init(mode="disabled")

In [25]:

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score,roc_auc_score
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

def compute_metrics(eval_pred):

    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    metrics = {}

    metrics['Accuracy'] = accuracy_score(labels, predictions)
    metrics['Precision'] = precision_score(labels, predictions)
    metrics['Recall'] = recall_score(labels, predictions)
    metrics['f1'] = f1_score(labels, predictions)
    metrics['AUC'] = roc_auc_score(labels, predictions)

    return(metrics)



Revision de parametros de TRAINER
https://huggingface.co/docs/transformers/main_classes/trainer




In [26]:
from transformers import TrainingArguments, Trainer

training_args = TrainingArguments(
    output_dir="beto_ejemplo_clase",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=2,
    weight_decay=0.01,
    eval_strategy="epoch",
    save_strategy="epoch",
    report_to="none",
    load_best_model_at_end=True,
    push_to_hub=True,
)

In [27]:

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train,
    eval_dataset=tokenized_test,
    processing_class=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)



In [28]:
# Train pre-trained model
trainer.train()

TrainOutput(global_step=1250, training_loss=0.29334074096679685, metrics={'train_runtime': 379.7042, 'train_samples_per_second': 52.673, 'train_steps_per_second': 3.292, 'total_flos': 598668095369280.0, 'train_loss': 0.29334074096679685, 'epoch': 2.0})

¿Cuanto tiempo les tardó el modelo?
¿CUantos EPOCH hizo? ¿Qué creen que es un epoch?

In [29]:
trainer.evaluate()

{'eval_loss': 0.31224334239959717,
 'eval_Accuracy': 0.8818654676893238,
 'eval_Precision': 0.6792156862745098,
 'eval_Recall': 0.4819143016138008,
 'eval_f1': 0.5638020833333334,
 'eval_AUC': 0.7195345654308266,
 'eval_runtime': 34.4202,
 'eval_samples_per_second': 329.545,
 'eval_steps_per_second': 20.598,
 'epoch': 2.0}

Esta fue la evaluación del entrenamiento que realicé yo usando todos los datos y no una muestra. ¿que opinan?



{'eval_loss': 0.6295255422592163,

 'eval_Accuracy': 0.891563078550648,

 'eval_Precision': 0.6768558951965066,

 'eval_Recall': 0.6037840845854201,

 'eval_f1': 0.6382352941176471,

 'eval_AUC': 0.7747602593469737,

 'eval_runtime': 36.0436,

 'eval_samples_per_second': 314.702,

 'eval_steps_per_second': 19.671,

 'epoch': 5.0}

Vayan a su cuenta de Hugginface y vean si está el modelo.




In [30]:
from transformers import pipelines

In [31]:
examples = ["hay que matarlos a todos", "Vamos a la plaza a marchar","Los chinos son todos sucios","Culpa de los bolitas"]


In [34]:
from transformers import pipeline # Import the pipeline function
examples = ["hay que matarlos a todos", "Vamos a la plaza a marchar","Los chinos son todos sucios","Culpa de los bolitas"]
# Specify the 'text-classification' task explicitly
specific_model = pipeline(task="text-classification", model="natdebandi/beto_ejemplo_clase7")
specific_model(examples)

model.safetensors:   0%|          | 0.00/439M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.27k [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/248k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/735k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/695 [00:00<?, ?B/s]

Device set to use cuda:0


[{'label': 'LABEL_1', 'score': 0.9847145676612854},
 {'label': 'LABEL_0', 'score': 0.9297218918800354},
 {'label': 'LABEL_1', 'score': 0.977299690246582},
 {'label': 'LABEL_1', 'score': 0.9890073537826538}]